packages/astro/src/runtime/client/hmr.ts

```diff
@@ -5,19 +5,38 @@ if (import.meta.hot) {
 		const { default: diff } = await import('micromorph');
 		const html = await fetch(`${window.location}`).then((res) => res.text());
 		const doc = parser.parseFromString(html, 'text/html');
-
+		for (const style of sheetsMap.values()) {
+			doc.head.appendChild(style);
+		}
 		// Match incoming islands to current state
 		for (const root of doc.querySelectorAll('astro-root')) {
 			const uid = root.getAttribute('uid');
 			const current = document.querySelector(`astro-root[uid="${uid}"]`);
 			if (current) {
-				root.innerHTML = current?.innerHTML;
+				current.setAttribute('data-persist', '');
+				root.replaceWith(current);
 			}
 		}
-		return diff(document, doc);
+		// both Vite and Astro's HMR scripts include `type="text/css"` on injected
+		// <style> blocks. These style blocks would not have been rendered in Astro's
+		// build and need to be persisted when diffing HTML changes.
+		for (const style of document.querySelectorAll("style[type='text/css']")) {
+			style.setAttribute('data-persist', '');
+			doc.head.appendChild(style.cloneNode(true));
```

Comment:
This is the fix for the Svelte component bug

Any `<style>` blocks that were injected by Astro or Vite as part of HMR should be persisted during the DOM diffing update